Src/FluentAssertions/Specialized/ExceptionAssertions.cs

```diff
@@ -193,7 +193,7 @@ public ExceptionAssertions(IEnumerable<TException> exceptions)
         return this;
     }
 
-    private Exception[] AssertInnerExceptionExactly(Type innerException, string because = "",
+    private IEnumerable<Exception> AssertInnerExceptionExactly(Type innerException, string because = "",
```

Comment:
Actually, for private methods there could be good reasons for using `IEnumerable<T>`, so I would only want this rule to kick in for `public`, `internal` and `protected` members.

BTW, according to the docs `CA1859` is disabled by default. How come you ran into that?

Reply:
> Actually, for private methods there could be good reasons for using `IEnumerable<T>`, so I would only want this rule to kick in for `public`, `internal` and `protected` members.

By "this rule" do you refer to Qodana or CA1859 here?
If we disable `ReturnTypeCanBeEnumerable.Local` I believe Qodana and .NET 8 would align.
 
> BTW, according to the docs `CA1859` is disabled by default. How come you ran into that?

https://github.com/fluentassertions/fluentassertions/blob/eb23c36ddb0841c67fe797af30f4397ccce21b58/Directory.Build.props#L16 includes CA1859 in the analysis.

Reply:
> I would only want this rule to kick in for public, internal and protected members.

[https://github.com/dotnet/roslyn-analyzers/issues/6564
Apperently this kicks in at compile time, but is disabled ](https://github.com/dotnet/roslyn-analyzers/issues/6564#issuecomment-1492985069)

Reply:
> > includes CA1859 in the analysis.
> 
> Hmm, that's scary. That means you get different results depending on whether you have the .NET 8 SDK installed?

`<AnalysisLevel>latest</AnalysisLevel>` is opt-in to all new "warning waves of the used compiler. 
As I understand the [documentation](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings#warninglevel) `<AnalysisLevel>4</AnalysisLevel>` or unspecified means use the warning wave released with TFM used for the project.

Reply:
But you're saying that if you install the .NET 8 SDK, it suddently starts to raise them? Or did you manually bump the analysis level or bump the TFM version?

Reply:
For this example using a .NET 8 compiler
```cs
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApp52;

static class A
{
    public static int Count() => Foo().Count();

    private static IEnumerable<int> Foo() => new int[] { 42 };
}
```

These three do not trigger CA1859:
```csproj
<TargetFramework>net7.0</TargetFramework>`
```

```csproj
<TargetFramework>net7.0</TargetFramework>
<AnalysisLevel>7.0</AnalysisLevel>
```

```csproj
<TargetFramework>net8.0</TargetFramework>
<AnalysisLevel>7.0</AnalysisLevel>
```

While when either having an unbounded analysis level (aka give me everything you got mr. Compiler) or bumping to .NET 8 without having a fixed analysis level, CA1859 is triggered.
Triggers CA1859:
```csproj
<TargetFramework>net7.0</TargetFramework>
<AnalysisLevel>latest</AnalysisLevel>
```

```csproj
<TargetFramework>net8.0</TargetFramework>
```

Perhaps we should fix the analysis level to 7.0 until we raise the required SDK to .NET 8?

Reply:
> Perhaps we should fix the analysis level to 7.0 until we raise the required SDK to .NET 8?

I would remove the `<AnalysisLevel>` property completely. Then, if I understand you correctly, it'll use the level that belongs to the target framework. I don't want people to be faced with unexpected warnings just because they have a newer SDK installed.

Reply:
We currently require the NET7 SDK to use C#11 features. This is specified in `global.json`.

Reply:
That doesn't conflict with what I'm saying, is it?

Reply:
> That doesn't conflict with what I'm saying, is it?

Not sure.

Removing `<AnalysisLevel>` and because the newest TFM we target is .NET 6, this enables SDK analyzers up to and including warning wave 6, released with the .NET 6 SDK.
Changing `<AnalysisLevel>` from `latest` to `7.0` also includes warning wave 7 released with the .NET 7 SDK, which is the minimum SDK required to compile Fluent Assertions.

So both ways stop new analyzers from kicking in when upgrading to a newer compiler than what is required, but fixing it to `7.0` also enables warning wave 7 analyzers which will be available due to our .NET 7 SDK requirement.

Reply:
> So both ways stop new analyzers from kicking in when upgrading to a newer compiler than what is required, but fixing it to `7.0` also enables warning wave 7 analyzers which will be available due to our .NET 7 SDK requirement.

OK, then it's clear to me. Let's fix it to `7.0` then.